### DISCUSSION TITLE
Goals & Objectives for clients (Roku, Android TV, etc.)?

### DISCUSSION THREAD

**Discussion by hoboken411** [jellyfin-archive/jellyfin-roku-legacy#915] [General]
Hi All - Big fan of the Jellyfin products. Excellent work across the board!
I've been scouring all the documentation - and I cannot seem to find the overall "scope" or "mission plan" for the product in general.

I have both the Android TV and Roku clients - and I assume timelines and technologies vary. However, is there a plan to get all clients to have the same "look and feel?"

For example, the Android TV client is extremely polished and functional compared to the Roku client.

Can anyone care to explain, in general terms, what your roadmap / timeline is for the product overall? In other words - is it your overall objective to get the Roku client to look the same as the Android TV client? Or what kind of challenges & limitations does each development platform present? Or why some platforms appear to have more contributors than others?

Sorry if this had been addressed elsewhere - but I'm curious if there is some kind of vision outlined somewhere.

Thanks again!

**Comment by cewert** [comment]
> Or what kind of challenges & limitations does each development platform present?

Roku uses it's own language called [BrightScript](https://developer.roku.com/en-gb/docs/references/brightscript/language/brightscript-language-reference.md). All features for the Roku client have to be custom coded or in other words we can't copy paste code from other clients. Roku doesn't use CSS so none of the custom themes for the web-client will work. The entire app has to be [less than 4MB](https://developer.roku.com/en-gb/docs/developer-program/getting-started/developer-setup.md#zip-files-and-packaging) after it's compressed. There are also certain components that all Roku applications must use such as the [built in video player](https://developer.roku.com/en-gb/docs/developer-program/core-concepts/playing-videos.md). Then there are hardware limitations - [not all rokus are very powerful](https://developer.roku.com/en-gb/docs/specs/hardware.md). specifically the roku sticks. And then there are weird bugs/quirks with Roku that never get documented like this [recent issue we ran into](https://github.com/jellyfin/jellyfin-roku/issues/1083) using the newest code features that roku provides us. Or [this weird bug](https://github.com/jellyfin/jellyfin-roku/issues/560) that caused the app to crash on startup ~10% of the time because of something undocumented. Those are the ones that come to mind there may be more I forget. Roku is kind of like a black box of mystery where you only know for sure how it works after you test it. Oh! and you can only test your code on a real roku attached to a monitor/tv. There is no emulator.

All of the above reasons are why the roku team has always been a small group. Not many people want to waste their time learning a new language that they can only use to make Roku apps. On top of that, the dev experience is crappy because of how you have to test your code and it's hard to find all the answers on google. This is also why the roku app has never had the best UI - it's hard enough to get stuff to work right let alone make it look pretty. We all want it to look pretty. It just ends up on the bottom of the priority list most of the time

**Reply by wilberfan** [reply]
An interesting discussion.   As someone who has been using Emby for quite awhile, I recently installed Jellyfin to give it a try after reading encouraging things about it.   I reluctantly switched back to Emby after discovering that specifically  the Jellyfin Roku app had fewer features than I was used to.  

I've had my Roku device for 3 years or so now, if I was interesting in switching to a different device, what would you recommend in terms of getting the most out of a current (or future) Jellyfin app?  An Nvidia Shield, a Chromecast, Apple TV, or...?  Thanks!

**Reply by cewert** [reply]
@wilberfan I have no idea. I only use the Roku client.

**Reply by joshuaboniface** [reply]
@wilberfan For the absolute best experience I tend to recommend something running AndroidTV; that app is by far the best supported and most feature-full, though Roku here is catching up fast as well and it's a toss-up between the two. If you do decide to go AndroidTV, FireTV sticks are relatively cheap (if you can stomach the endless Amazon ads).

**Reply by hoboken411** [reply]
@cewert Thanks so much for the detailed reply. I understand better now - that the platforms are really different, and that it's not as simple as you might think. It seems like Roku is stuck in a bind really - having to support every device. I wonder if, at some point, the way smartphones evolve and "drop support" for older models - that Roku will eventually do the same. A conundrum considering the size (and relative non-techy aspect) of their user group.

**Reply by cewert** [reply]
> I wonder if, at some point, the way smartphones evolve and "drop support" for older models - that Roku will eventually do the same.

@hoboken411 That's how it already works. If you look at the [legacy models](https://developer.roku.com/en-gb/docs/specs/hardware.md#legacy-models) you'll notice the only reason they are "discontinued" is because they are too old to run the newer versions of Roku OS. And just like Android, developers can choose which version of the Roku OS (aka API) to use. So if an android or roku app uses the latest version of the OS then the devices who can't run that version of the OS won't be able to run the app.

So, in theory, you could make a Roku app that supports all roku devices, even the legacy models, if you code your app using Roku OS 3.1 or below. In practice, that's impossible because Roku has deprecated some essential bits of the API that you'd need to use. And that continues to happen as time goes by - Roku will deprecate certain parts of the API and all Roku apps have to adjust.
